### PR TITLE
[NN] Refactor DegreeEncoder, SpatialEncoder and PathEncoder

### PR DESCRIPTION
**Pull request by ZHITENGLI**
## Description

To improve the efficiency of DegreeEncoder, SpatialEncoder and PathEncoder, we'd like to separate the padding operations from these modules. Specifically, the padding operations will be implemented in the `collate` function and the input graph will be replaced by padded features.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [x] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### CODE REVIEW

File: python/dgl/nn/pytorch/gt/degree_encoder.py [hunk 1]
```diff
@@ -31,10 +29,17 @@ class DegreeEncoder(nn.Module):
...
+    >>> from torch.nn.utils.rnn import pad_sequence
 
-    >>> g = dgl.graph(([0,0,0,1,1,2,3,3], [1,2,3,0,3,0,0,1]))
+    >>> g1 = dgl.graph(([0,0,0,1,1,2,3,3], [1,2,3,0,3,0,0,1]))
+    >>> g2 = dgl.graph(([0,1], [1,0]))
+    >>> in_degree = pad_sequence([g1.in_degrees(), g2.in_degrees()], batch_first=True)
+    >>> out_degree = pad_sequence([g1.out_degrees(), g2.out_degrees()], batch_first=True)
     >>> degree_encoder = DegreeEncoder(5, 16)
-    >>> degree_embedding = degree_encoder(g)
+    >>> degree_embedding = degree_encoder(th.stack((in_degree, out_degree)))
```
Comment (rudongyu) [inline python/dgl/nn/pytorch/gt/degree_encoder.py:40]: maybe print the input of `degree_encoder` first for those who are not familiar with `pad_sequence`

File: python/dgl/nn/pytorch/gt/degree_encoder.py [hunk 2]
```diff
@@ -53,36 +58,32 @@ def __init__(self, max_degree, embedding_dim, direction="both"):
...
         ----------
-        g : DGLGraph
-            A DGLGraph to be encoded. Graphs with more than one type of edges
-            are not allowed.
+        degrees : Tensor
+            If direction is ``both``, it should be the stack of in degrees and out degrees
+            of the batched graph with zero padding, of shape :math:`(2, B, N)`.
+            Otherwise, it can be the in degrees or out degrees of the batched graph
+            with zero padding, of shape :math:`(B, N)`, where :math:`B` is the
+            batch size of the batched graph, and :math:`N` is the maximum number of nodes.
```
Comment (rudongyu) [inline python/dgl/nn/pytorch/gt/degree_encoder.py:70]:
```suggestion
            If :attr:`direction` is ``both``, it should be stacked in degrees and out degrees
            of the batched graph with zero padding, a tensor of shape :math:`(2, B, N)`.
            Otherwise, it should be zero-padded in degrees or out degrees of the batched
            graph, a tensor of shape :math:`(B, N)`, where :math:`B` is the
            batch size of the batched graph, and :math:`N` is the maximum number of nodes.
```

File: python/dgl/nn/pytorch/gt/degree_encoder.py [hunk 3]
```diff
@@ -53,36 +58,32 @@ def __init__(self, max_degree, embedding_dim, direction="both"):
...
+            where :math:`d` is :attr:`embedding_dim`.
         """
-        if len(g.etypes) > 1:
-            raise DGLError(
-                "The input graph should have no more than one type of edges."
-            )
-
-        in_degree = th.clamp(g.in_degrees(), min=0, max=self.max_degree)
-        out_degree = th.clamp(g.out_degrees(), min=0, max=self.max_degree)
+        degrees = th.clamp(degrees, min=0, max=self.max_degree)
```
Comment (rudongyu) [inline python/dgl/nn/pytorch/gt/degree_encoder.py:79]: Should we include a check for dimensions of `degrees` in each branch below to avoid silent incorrect encoding?

File: python/dgl/nn/pytorch/gt/path_encoder.py [hunk 4]
```diff
@@ -31,13 +28,19 @@ class PathEncoder(nn.Module):
...
     >>> import dgl
     >>> from dgl.nn import PathEncoder
+    >>> from dgl import shortest_dist
 
-    >>> u = th.tensor([0, 0, 0, 1, 1, 2, 3, 3])
-    >>> v = th.tensor([1, 2, 3, 0, 3, 0, 0, 1])
-    >>> g = dgl.graph((u, v))
+    >>> g = dgl.graph(([0,0,0,1,1,2,3,3], [1,2,3,0,3,0,0,1]))
     >>> edata = th.rand(8, 16)
+    >>> edata = th.cat(
```
Comment (rudongyu) [inline python/dgl/nn/pytorch/gt/path_encoder.py:35]: add a comment here to describe why we need zero paddings in edata (-1 paddings in `path` tensor)

File: python/dgl/nn/pytorch/gt/spatial_encoder.py [hunk 5]
```diff
@@ -33,14 +32,18 @@ class SpatialEncoder(nn.Module):
...
-
-    >>> u = th.tensor([0, 0, 0, 1, 1, 2, 3, 3])
-    >>> v = th.tensor([1, 2, 3, 0, 3, 0, 0, 1])
-    >>> g = dgl.graph((u, v))
+    >>> from dgl import shortest_dist
+
+    >>> g1 = dgl.graph(([0,0,0,1,1,2,3,3], [1,2,3,0,3,0,0,1]))
+    >>> g2 = dgl.graph(([0,1], [1,0]))
+    >>> n1, n2 = g1.num_nodes(), g2.num_nodes()
+    >>> dist = -th.ones((2, 4, 4), dtype=th.long)
```
Comment (rudongyu) [inline python/dgl/nn/pytorch/gt/spatial_encoder.py:40]: ditto. explain the "-1" paddings with a comment.

File: python/dgl/nn/pytorch/gt/spatial_encoder.py [hunk 6]
```diff
@@ -52,41 +55,29 @@ def __init__(self, max_dist, num_heads=1):
...
-    def forward(self, g):
+    def forward(self, dist):
         """
         Parameters
         ----------
-        g : DGLGraph
-            A DGLGraph to be encoded, which must be a homogeneous one.
+        dist : Tensor
+            Shortest path distance matrix of the batched graph with -1 padding,
+            of shape :math:`(B, N, N)`, where :math:`B` is the batch size of
```
Comment (rudongyu) [inline python/dgl/nn/pytorch/gt/spatial_encoder.py:64]:
```suggestion
            Shortest path distance of the batched graph with -1 padding,
            a tensor of shape :math:`(B, N, N)`, where :math:`B` is the batch size of
```

**Comment by rudongyu** [comment]
@dgl-bot

**Comment by rudongyu** [comment]
@dgl-bot

**Review by rudongyu — APPROVED** [review]
